experimental/serde_unaligned/src/uvec/mod.rs

```diff
@@ -0,0 +1,321 @@
+where
+    for<'h> T: Copy + From<ByteSliceLE<'h, 4>> + PartialEq,
+{
+    fn eq(&self, other: &UVec<'b, T>) -> bool {
+        if self.len() != other.len() {
+            return false;
+        }
+        // TODO: Use an iterator here
+        for i in 0..self.len() {
+            if self.get(i) != other.get(i) {
+                return false;
+            }
+        }
+        true
+    }
+}
+
+impl<'a, T> UVec<'a, T> {
+    pub fn from_unaligned_le_bytes(bytes: &'a [u8]) -> Self {
+        Self(UVecInner::UnalignedLE(bytes))
+    }
+
+    pub fn from_unaligned_ne_bytes(_bytes: &'a [u8]) -> Self {
+        unimplemented!()
+    }
+
+    pub fn len(&self) -> usize {
+        use UVecInner::*;
+        match &self.0 {
+            Owned(vec) => vec.len(),
+            Aligned(slice) => slice.len(),
+            UnalignedLE(bytes) => bytes.len() / std::mem::size_of::<T>(),
+        }
+    }
+
+    pub fn into_inner(self) -> UVecInner<'a, T> {
+        self.0
+    }
+}
+
+// TODO: I guess we can't implement std::ops::Index because that returns references, right?
+impl<'a, T> UVec<'a, T>
+where
+    for<'h> T: Copy + From<ByteSliceLE<'h, 4>>,
+{
+    pub fn get(&self, index: usize) -> Option<T> {
+        use UVecInner::*;
+        match &self.0 {
+            Owned(vec) => vec.get(index).copied(),
+            Aligned(slice) => slice.get(index).copied(),
+            UnalignedLE(bytes) => bytes
+                .as_chunks::<4>()
+                .0
+                .get(index)
+                .map(ByteSliceLE::<4>::from)
+                .map(T::from),
+        }
+    }
+}
+
+impl<'a, T> UVec<'a, T>
+where
+    T: Copy,
+    ByteArrayLE<4>: From<T>,
+{
+    pub fn write_to_stream_le(&self, stream: &mut dyn std::io::Write) -> std::io::Result<usize> {
+        use UVecInner::*;
+        let mut result: usize = 0;
+        match &self.0 {
+            Owned(vec) => {
+                for value in vec.iter() {
+                    result += stream.write(&ByteArrayLE::<4>::from(*value).as_bytes())?;
+                }
+            }
+            Aligned(slice) => {
+                for value in slice.iter() {
+                    result += stream.write(&ByteArrayLE::<4>::from(*value).as_bytes())?;
+                }
+            }
+            UnalignedLE(bytes) => {
+                result += stream.write(bytes)?;
+            }
+        };
+        Ok(result)
+    }
+
+    pub fn to_le_bytes(&self) -> Cow<'a, [u8]> {
+        use UVecInner::*;
+        if let UnalignedLE(bytes) = self.0 {
+            return Cow::Borrowed(bytes);
+        }
+        let mut bytes: Vec<u8> = Vec::with_capacity(self.len() * std::mem::size_of::<T>());
+        self.write_to_stream_le(&mut bytes)
+            .expect("Writing to memory buffer");
+        Cow::Owned(bytes)
+    }
+}
+
+impl<'a, T> UVec<'a, T>
+where
+    T: Copy + Default + std::ops::AddAssign + From<ByteSliceLE<'a, 4>>,
+{
+    pub fn sum(&self) -> T {
```

Comment:
I would just implement a moving iterator instead of doing this, all of these things can be handled by Iterator

Reply:
Agreed; I'll do that next.